src/sql/engine/sort/ob_sort_op_impl.h

```diff
@@ -436,6 +441,40 @@ class ObSortOpImpl
     { get_extra_info().max_size_ = max_size; }
   };
 
+  class SelfSharpFilter
+  {
+  public:
+    static const int64_t MAX_BUCKET_NUM = 100;
+    static const int64_t MIN_BUCKET_NUM = 10;
+    static constexpr double FILTER_MEM_RATIO  = 0.05; // limit the ratio of memory used
+
+    typedef common::ObBinaryHeap<ObChunkDatumStore::StoredRow *, Compare, MAX_BUCKET_NUM> Filter;
+
+    SelfSharpFilter();
+
+    int init(lib::MemoryContext &mem_context, Compare &comp);
+    void cal_hyper_params(const uint64_t rows_limit, const int64_t topn_cnt);
+    bool is_inited() const { return heap_ != NULL && per_bucket_num_ != 0 && bucket_num_ != 0; }
+    // filter single data
+    int filter(const common::ObIArray<ObExpr*> *exprs, Compare &comp, ObEvalCtx &eval_ctx, bool &ignore);
+    // filter batch data
+    int filter(const common::ObIArray<ObExpr*> *exprs, ObBitVector &skip, const int64_t batch_size,
+               const int64_t start_pos, Compare &comp, ObEvalCtx &eval_ctx);
+    int update_filter(ObChunkDatumStore::StoredRow *new_row, lib::MemoryContext &mem_context,
+                      ObSqlMemMgrProcessor &sql_mem_processor, Compare &comp,
+                      int64_t &inmem_row_size, uint64_t seq_no, bool use_heap_sort);
+
+    void reset(lib::MemoryContext &mem_context);
+    void reuse(lib::MemoryContext &mem_context, ObSqlMemMgrProcessor &sql_mem_processor);
+    inline uint64_t per_bucket_num() { return per_bucket_num_; }
+
+  private:
+    Filter *heap_;            // heap stored statistics info
+    uint64_t per_bucket_num_; // numb of rows in each bucket
+    uint64_t bucket_num_;     // num of buckets
+    bool try_cal_params_;     // initialization was attempted
```

Comment:
Not needed? bucket_num_ may be used to indicate inited or not.

Reply:
This bool variable ensures that the SelfSharpFilter calculates the bucket size only once.

If you find on the first attempt that there is no need to enable filter optimization, then there is no need to try again.